.generator/test_cli.py

```diff
@@ -807,3 +830,128 @@ def test_process_version_file_failure():
+            "google.cloud",
+        ),
+        (
+            "repo/packages/google-ads",
+            "repo/packages/google-ads/google/ads/v17",
+            "google.ads",
+        ),
+        (
+            "repo/packages/google-auth",
+            "repo/packages/google-auth/google/auth",
+            "google",
+        ),
+        ("repo/packages/google-api", "repo/packages/google-api/google", "google"),
+    ],
+)
+def test_determine_library_namespace_success(
+    pkg_root_str, gapic_parent_str, expected_namespace
+):
+    """Tests that the refactored namespace logic correctly calculates the relative namespace."""
+    pkg_root_path = Path(pkg_root_str)
+    gapic_parent_path = Path(gapic_parent_str)
+
+    namespace = _determine_library_namespace(gapic_parent_path, pkg_root_path)
+    assert namespace == expected_namespace
+
+
+def test_determine_library_namespace_fails_not_subpath():
+    """Tests that a ValueError is raised if the gapic path is not inside the package root."""
+    pkg_root_path = Path("repo/packages/my-lib")
+    gapic_parent_path = Path("SOME/OTHER/PATH/google/cloud/api")
+
+    with pytest.raises(ValueError):
+        _determine_library_namespace(gapic_parent_path, pkg_root_path)
+
+
+def test_verify_library_namespace_success_valid(mocker, mock_path_class):
+    """Tests success when a single valid namespace is found."""
+    # 1. Get the mock instance from the mock class's return_value
+    mock_instance = mock_path_class.return_value
+
+    # 2. Configure the mock instance
+    mock_instance.is_dir.return_value = True
+    mock_file = MagicMock(spec=Path)
+    mock_file.parent = Path("/abs/repo/packages/my-lib/google/cloud/language")
+    mock_instance.rglob.return_value = [mock_file]
+
+    mock_determine_ns = mocker.patch(
+        "cli._determine_library_namespace", return_value="google.cloud"
+    )
+
+    result = _verify_library_namespace("my-lib", "/abs/repo")
+
+    assert result is True
+
+    # 3. Assert against the mock CLASS (from the fixture)
+    mock_path_class.assert_called_once_with("/abs/repo/packages/my-lib")
+
+    # 4. Verify the helper was called with the correct instance
+    mock_determine_ns.assert_called_once_with(mock_file.parent, mock_instance)
+
+
+def test_verify_library_namespace_failure_invalid(mocker, mock_path_class):
+    """Tests failure when a namespace is found that is NOT in the valid list."""
+    mock_instance = mock_path_class.return_value
+    mock_instance.is_dir.return_value = True
+
+    mock_file = MagicMock(spec=Path)
+    mock_file.parent = Path("/abs/repo/packages/my-lib/google/api/core")
+    mock_instance.rglob.return_value = [mock_file]
+
+    mock_determine_ns = mocker.patch(
+        "cli._determine_library_namespace", return_value="google.api"
+    )
+
+    expected_error = "The namespace `google.api` for `my-lib` must be one of ['google', 'google.cloud', 'google.ads', 'google.cloud.billing']."
```

Comment:
nit: remove these so that we don't need to manually maintain this text when we add new registered namespaces